core/auth_provider_okta.go

```
@@ -0,0 +1,311 @@
+
+	req, _ := http.NewRequest("POST", url, bytes.NewReader([]byte("")))
+	h := req.Header
+	h.Add("Authorization", "Basic "+authHeader)
+	h.Add("Accept", "application/json")
+	h.Add("Content-Type", "application/x-www-form-urlencoded")
+	h.Add("Connection", "close")
+	h.Add("Content-Length", "0")
+
+	client := &http.Client{}
+	resp, _ := client.Do(req)
+	body, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	var exchange Exchange
+	json.Unmarshal(body, &exchange)
+
+	if exchange.Error != "" {
+		p.Errorf("exchange resulted in error %s: %s", exchange.Error, exchange.ErrorDescription)
+		return nil
+	}
+
+	// Now we succussfully exhchanged the auth code for an openID token which is stored in our struct "exchange"
+	// Next we verify the token(s) and extract our okta claims from it (i.e. groups)
+	// Goal is to get {name, account, and orgs} from this token to do the mapping. 
+
+	id_token, verificationError := p.verifyToken(exchange.IdToken, "id")
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+	access_token, verificationError := p.verifyToken(exchange.AccessToken, "access")
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+
+	groups_claim := id_token.Claims["groups"]
+	if groups_claim == nil{
+		groups_claim = access_token.Claims["groups"]  		//ex: access_token.groups_claim - [[Everyone User Admin]]
+	}
+	// log.Debugf("DEBUG::groups_claim [%s]", groups_claim)
+
+
+	// TODO: get orgs from okta endpoint
+	org_names := make([]string, 0)
+	org_names = append(org_names,"okta")
+
+	orgs := make(map[string][]string)
+	for _, org_name := range org_names {
+		orgs[org_name] = make([]string, 0)
+	}														//ex: orgs- [map[okta:[]]] - [map[string][]string]
+
+	g, err := json.Marshal(groups_claim) 					//ex: g-[["Everyone","User","Admin"]] - [[]uint8]
+	if err != nil {
+		fmt.Println(err)
+	}
+	groups_list :=  make([]string, 0)
+	json.Unmarshal(g, &groups_list)							//ex: groups_list-[[Everyone User Admin]] - [[]string]
```

Do not silence errors.